platform/linux-generic/odp_ipsec.c

```diff
@@ -565,18 +593,40 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
+	if (ipsec_sa->esn)
+		ipsec_sa->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
+
+	/* AAD with ESN */
+	if (ipsec_sa->esn && (!ipsec_sa->seqh_len))
+		state->esp.aad.seq_no64 = odp_cpu_to_be_64(ipsec_sa->in.seq64);
+	else if (!ipsec_sa->esn)
+		state->esp.aad.seq_no = esp.seq_no;
 
 	param->aad_ptr = (uint8_t *)&state->esp.aad;
 
+	/* Inbound ICV computation includes ESN higher 32 bits for individual algo's */
+	if (ipsec_sa->seqh_len) {
+		uint32_t inb_seqh = (ipsec_sa->in.seq64 >> 32);
+		uint32_t icv_offset =  odp_packet_len(*pkt) - ipsec_sa->icv_len;
+
+		if (odp_packet_add_data(pkt, icv_offset, ipsec_sa->seqh_len) < 0) {
+			status->error.alg = 1;
+			ODP_ERR("odp_packet_add_data failed\n");
+			return -1;
+		}
+		odp_packet_copy_from_mem(*pkt, icv_offset, ipsec_sa->seqh_len, &inb_seqh);
+	}
+
 	param->auth_range.offset = ipsec_offset;
 	param->auth_range.length = state->ip_tot_len -
 				  state->ip_hdr_len -
-				  ipsec_sa->icv_len;
+				  ipsec_sa->icv_len +
+				  ipsec_sa->seqh_len;
 	param->hash_result_offset = state->ip_offset +
 				   state->ip_tot_len -
-				   ipsec_sa->icv_len;
+				   ipsec_sa->icv_len +
+				   ipsec_sa->seqh_len;
```

Comment:
Please keep the summation order the same as the order of the fields in the packet in both the offset and the length calculation).

Reply:
Ack, will fix it.